feature/better-amount-input

DataCourier opened the pull request:
Work in progress, seeking comments right now.
- [x] finish test suite
- [x] simplify code
- [x] add some comments
- [x] get feedback
- [x] apply component to the other form as well
- [x] branch needs to be renamed
- [x] squash commits and commit with nice messages
- [ ] <del>modify to support mili-satoshis</del>

![feb-11-2018 14-44-55](https://user-images.githubusercontent.com/35670446/36078976-15a12b42-0f7e-11e8-80c7-b84cd5336119.gif)

Empact commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,134 @@
+      integer = e.target.value * 1;
+    }
+    if (value.match(/[0-9]*[,\.][0-9]+/)) {
+      [integer, fractional] = value.toString().split('.');
```
The split could be a `.` or ` `,` as per the regex.

DataCourier replied:
fixed

Empact commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,134 @@
+    if (value.match(/[0-9]+/)) {
+      integer = e.target.value * 1;
+    }
+    if (value.match(/[0-9]*[,\.][0-9]+/)) {
```
No need to escape the `.` inside a character class.

DataCourier replied:
fixed

Empact commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // limit fractional precision to 8 places
+    if (fractional && fractional.length > 8) {
```
Limit is 11 characters for off-chain, 8 for on-chain.
https://github.com/lightningnetwork/lnd/blob/master/lnwire/msat.go

Empact replied:
This is incorrect - satoshis are still the unit for off-chain payments. Higher precision is used to accommodate fees.

JimmyMow commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,173 @@
+import styles from './RequestForm.scss'
+
+class AmountInput extends React.Component {
+
```
no need for empty line here

JimmyMow commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  formatValue(integer, fractional) {
+    let value;
```
our code style is no semicolons for now :p

JimmyMow commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,173 @@
+  currency: PropTypes.string.isRequired,
+  crypto: PropTypes.string.isRequired,
+  onChange:PropTypes.func.isRequired,
+  amount: PropTypes.string.isRequired,
```
can remove the trailing comma

JimmyMow commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,173 @@
+    // extract the parsing bit?
+    if (value.match(/^[0-9]*[.,][0-9]*$/)) {
+      [integer, fractional] = value.toString().split(/[.,]/)
+      if (!fractional || fractional == '') {
```
hmm, is this conditional needed? if `fractional` evaluates to false or `fractional == ''` then the conditional below will evaluate to false anyways right?

JimmyMow commented on app/components/Form/AmountInput.js:
```diff
@@ -0,0 +1,173 @@
+    if (fractional && fractional.length > 0) {
+      value = `${integer}.${fractional}`
+    } else { 
+      if (fractional === '') {
```
this conditional sets value = `${integer}.` no matter what. doesn't look like it's needed

DataCourier replied:
When integer is null, it runs the else part. Thats how it knows its just `X.` vs just `X`

JimmyMow commented on app/components/Form/RequestForm.js:
```diff
@@ -2,6 +2,7 @@ import React from 'react'
 import PropTypes from 'prop-types'
 import CurrencyIcon from 'components/CurrencyIcon'
 import styles from './RequestForm.scss'
+import AmountInput from './AmountInput'
```
Should be imported above the `styles`

JimmyMow commented:
Nice @DataCourier! This is an overall improvement. I left some small comments throughout, looks good overall. A few changes and passing travis and I'd love to merge this. 

Also, why did you not use this in the `PayForm`? Just curious if there was a specific reason

JimmyMow commented:
Also extracting out the `AmountInput` into its own component is a wonderful idea and a step in the right direction. I'm going to be adding multi-currencies support soon (BTC, bits, USD, etc). This component will be super helpful then as this form gets more complex

DataCourier commented:
@JimmyMow I'll be applying it in the other form as well (its in the checklist), just wanted to get general feedback on the behavior first. Will be addressing the comments over the weekend most likely.

DataCourier commented:
Renamed branch, closing PR for the new one https://github.com/LN-Zap/zap-desktop/pull/322
